File: android/src/main/kotlin/ch/waio/pro_video_editor/src/shared/logging/PluginLog.kt
package ch.waio.pro_video_editor.src.shared.logging

import android.util.Log as AndroidLog

object PluginLog {
    private const val LEVEL_NONE = Int.MAX_VALUE

    @Volatile
    private var minimumPriority = defaultPriority()

    /**
     * Optional sink that forwards every emitted log entry to Flutter.
     *
     * Set by the plugin while a Dart listener is attached to the
     * `pro_video_editor_logs` event channel. The level string matches
     * `NativeLogLevel.methodValue` on the Dart side.
     */
    @Volatile
    var sink: ((level: String, tag: String, message: String, throwable: Throwable?) -> Unit)? = null

    fun setMinimumLevel(level: String) {
        minimumPriority = parsePriority(level)
    }

    private fun defaultPriority(): Int {
        val isDebugBuild = runCatching {
            Class.forName("ch.waio.pro_video_editor.BuildConfig")
                .getField("DEBUG")
                .getBoolean(null)
        }.getOrDefault(true)

        return if (isDebugBuild) AndroidLog.DEBUG else AndroidLog.WARN
    }

    private fun parsePriority(level: String): Int {
        return when (level.lowercase()) {
            "none" -> LEVEL_NONE
            "error" -> AndroidLog.ERROR
            "warn", "warning" -> AndroidLog.WARN
            "info" -> AndroidLog.INFO
            "debug" -> AndroidLog.DEBUG
            "verbose" -> AndroidLog.VERBOSE
            else -> throw IllegalArgumentException("Unsupported Android log level: $level")
        }
    }

    private fun shouldLog(priority: Int): Boolean {
        return priority >= minimumPriority
    }

    private fun levelName(priority: Int): String {
        return when (priority) {
            AndroidLog.VERBOSE -> "verbose"
            AndroidLog.DEBUG -> "debug"
            AndroidLog.INFO -> "info"
            AndroidLog.WARN -> "warning"
            AndroidLog.ERROR -> "error"
            else -> "info"
        }
    }

    /**
     * Forwards an entry to the Dart [sink], if one is attached.
     *
     * Called only after the priority passed [shouldLog], so the Dart stream
     * is gated by the same level as the native console output.
     */
    private fun forward(priority: Int, tag: String, message: String, throwable: Throwable?) {
        sink?.invoke(levelName(priority), tag, message, throwable)
    }

    fun v(tag: String, message: String): Int {
        if (!shouldLog(AndroidLog.VERBOSE)) return 0
        forward(AndroidLog.VERBOSE, tag, message, null)
        return AndroidLog.v(tag, message)
    }

    fun d(tag: String, message: String): Int {
        if (!shouldLog(AndroidLog.DEBUG)) return 0
        forward(AndroidLog.DEBUG, tag, message, null)
        return AndroidLog.d(tag, message)
    }

    fun i(tag: String, message: String): Int {
        if (!shouldLog(AndroidLog.INFO)) return 0
        forward(AndroidLog.INFO, tag, message, null)
        return AndroidLog.i(tag, message)
    }

    fun w(tag: String, message: String): Int {
        if (!shouldLog(AndroidLog.WARN)) return 0
        forward(AndroidLog.WARN, tag, message, null)
        return AndroidLog.w(tag, message)
    }

    fun w(tag: String, message: String, throwable: Throwable): Int {
        if (!shouldLog(AndroidLog.WARN)) return 0
        forward(AndroidLog.WARN, tag, message, throwable)
        return AndroidLog.w(tag, message, throwable)
    }

    fun e(tag: String, message: String): Int {
        if (!shouldLog(AndroidLog.ERROR)) return 0
        forward(AndroidLog.ERROR, tag, message, null)
        return AndroidLog.e(tag, message)
    }

    fun e(tag: String, message: String, throwable: Throwable): Int {
        if (!shouldLog(AndroidLog.ERROR)) return 0
        forward(AndroidLog.ERROR, tag, message, throwable)
        return AndroidLog.e(tag, message, throwable)
    }
}

File: darwin/pro_video_editor/Sources/pro_video_editor/src/shared/logging/PluginLog.swift
import Foundation

enum PluginLogLevel: Int {
  case verbose = 0
  case debug = 1
  case info = 2
  case warning = 3
  case error = 4
  case none = 5

  static var `default`: PluginLogLevel {
    #if DEBUG
      return .debug
    #else
      return .warning
    #endif
  }

  /// String value matching `NativeLogLevel.methodValue` on the Dart side.
  var methodValue: String {
    switch self {
    case .verbose: return "verbose"
    case .debug: return "debug"
    case .info: return "info"
    case .warning: return "warning"
    case .error: return "error"
    case .none: return "none"
    }
  }

  static func from(methodValue: String) throws -> PluginLogLevel {
    switch methodValue.lowercased() {
    case "verbose":
      return .verbose
    case "debug":
      return .debug
    case "info":
      return .info
    case "warning", "warn":
      return .warning
    case "error":
      return .error
    case "none":
      return .none
    default:
      throw PluginLogError.unsupportedLevel(methodValue)
    }
  }
}

enum PluginLogError: LocalizedError {
  case unsupportedLevel(String)

  var errorDescription: String? {
    switch self {
    case .unsupportedLevel(let level):
      return "Unsupported native log level: \(level)"
    }
  }
}

enum PluginLog {
  private static var minimumLevel = PluginLogLevel.default

  /// Optional sink that forwards every emitted log entry to Flutter.
  ///
  /// Set by the plugin while a Dart listener is attached to the
  /// `pro_video_editor_logs` event channel.
  static var sink: ((_ level: PluginLogLevel, _ message: String) -> Void)?

  static func setMinimumLevel(_ methodValue: String) throws {
    minimumLevel = try .from(methodValue: methodValue)
  }

  static func print(_ items: Any..., separator: String = " ", terminator: String = "\n") {
    let message = items.map { String(describing: $0) }.joined(separator: separator)
    log(message, terminator: terminator)
  }

  private static func log(_ message: String, terminator: String) {
    let level = inferredLevel(for: message)
    guard shouldLog(level) else { return }
    Swift.print(message, terminator: terminator)
    sink?(level, message)
  }

  private static func shouldLog(_ level: PluginLogLevel) -> Bool {
    level.rawValue >= minimumLevel.rawValue && minimumLevel != .none
  }

  private static func inferredLevel(for message: String) -> PluginLogLevel {
    if message.contains("❌") || message.localizedCaseInsensitiveContains("error")
      || message.localizedCaseInsensitiveContains("failed")
    {
      return .error
    }

    if message.contains("⚠️") || message.localizedCaseInsensitiveContains("warn") {
      return .warning
    }

    if message.contains("ℹ️") || message.localizedCaseInsensitiveContains("info") {
      return .info
    }

    return .debug
  }
}
